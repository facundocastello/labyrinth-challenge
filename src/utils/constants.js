export const initSize = 5;
export const maxSize = 20;
export const initStartPosition = 0;
export const initEndPosition = initSize * initSize - 1;