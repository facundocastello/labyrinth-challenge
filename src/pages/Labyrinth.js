import React, { useCallback, useEffect, useState } from 'react'
import { ArrowLeft, CornerUpLeft, RefreshCw, Settings } from 'react-feather';
import Board from '../components/Board'
import Container from '../components/Container';
import GameTypography from '../components/GameTypography';
import Joystick from '../components/Joystick';
import Menu from '../components/Menu';
import Row from '../components/Row';
import Token from '../components/Token';
import { initEndPosition, initStartPosition, initSize, maxSize } from '../utils/constants';

export default function Labyrinth() {
  const [edittingStatus, setEdittingStatus] = useState(false);
  const [blockedPositions, setBlockedPositions] = useState([]);
  const [size, setSize] = useState(initSize);
  const [endPosition, setEndPosition] = useState(initEndPosition);
  const [startPosition, setStartPosition] = useState(initStartPosition);
  const [tokenPositions, setTokenPositions] = useState([startPosition]);
  const [limitMovements, setLimitMovements] = useState((2 * initSize) - 1);
  const [time, setTime] = useState(undefined);

  useEffect(() => {
    if (tokenPositions.length > 1 && !time) {
      setTime(new Date());
    } else if (endPosition === tokenPositions[0] && typeof time !== 'number') {
      setTime((new Date() - time) / 1000)
    }
  }, [endPosition, time, tokenPositions])

  useEffect(() => {
    if (tokenPositions.length === 1) setTime(undefined)
  }, [tokenPositions.length])

  const generatePath = useCallback((start, end, size) => {
    const path = [start];
    const xEnd = end % size
    const yEnd = Math.floor(end / size)
    const orientation = Math.random();
    while (path[path.length - 1] !== end) {
      let newBlock = path[path.length - 1];
      const xNewBlock = newBlock % size;
      const yNewBlock = Math.floor(newBlock / size);
      if (xEnd !== xNewBlock && (Math.random() > orientation || yNewBlock === yEnd)) {
        if (xNewBlock < xEnd) path.push(newBlock + 1)
        if (xNewBlock > xEnd) path.push(newBlock - 1)
      } else {
        if (yNewBlock < yEnd) path.push(newBlock + 1 * size)
        if (yNewBlock > yEnd) path.push(newBlock - 1 * size)
      }
    }
    return path;
  }, [])

  const setRandomBlocks = useCallback((startPosition, endPosition, size) => {
    const maxBlock = size * size;
    const happyPath = generatePath(startPosition, endPosition, size);
    setLimitMovements(happyPath.length)
    let falsePaths = [];
    for (let index = 0; index < Math.floor(size / 4); index++) {
      const start = happyPath[Math.round(Math.random() * (happyPath.length - 1))]
      const end = Math.round(Math.random() * maxBlock);
      falsePaths = [...falsePaths, ...generatePath(start, end, size)];
    }

    const newBlockedPositions = [];
    for (let x = 0; x < size; x++) {
      for (let y = 0; y < size; y++) {
        const position = x * size + y;
        if (!happyPath.includes(position) && !falsePaths.includes(position))
          newBlockedPositions.push(position);
      }
    }
    setBlockedPositions(newBlockedPositions)
    setTokenPositions([startPosition]);
  }, [generatePath])

  const scalarBaseChange = useCallback((position, oldScalar, newScalar) => {
    const x = position % oldScalar;
    const y = Math.floor(position / oldScalar);
    const newX = Math.round((x - oldScalar / 2) * newScalar / oldScalar + (newScalar/2));
    const newY = Math.round((y - oldScalar / 2) * newScalar / oldScalar + (newScalar/2));

    debugger;
    return newX + newY * newScalar;
  },[])

  const changeSize = useCallback((ev) => {
    const newSize = ev.target ? ev.target.value * 1 : ev;
    if (newSize > maxSize || newSize < initSize) return
    const newStartPosition = scalarBaseChange(startPosition, size, newSize);
    const newEndPosition = scalarBaseChange(endPosition, size, newSize);
    setSize(newSize);
    setEndPosition(newEndPosition);
    setTokenPositions([newStartPosition])
    setStartPosition(newStartPosition);
    setLimitMovements(Math.floor(2.5 * newSize));
    setRandomBlocks(newStartPosition, newEndPosition, newSize);
  }, [endPosition, scalarBaseChange, setRandomBlocks, size, startPosition]);

  const handleLimitMovementsChange = (ev) => {
    setLimitMovements(ev.target.value * 1 + 1);
  }

  useEffect(() => {
    setRandomBlocks(initStartPosition, initSize * initSize - 1, initSize);
  }, [setRandomBlocks])

  const handleSetRandom = useCallback(() => {
    setRandomBlocks(startPosition, endPosition, size);
  }, [endPosition, setRandomBlocks, size, startPosition])

  const handleSlotClicked = (position) => {
    switch (edittingStatus) {
      case 'block':
        const newBlockedPositions = [...blockedPositions];
        const blockIndex = newBlockedPositions.indexOf(position);
        if (blockIndex === -1) {
          newBlockedPositions.push(position);
        } else {
          newBlockedPositions.splice(blockIndex, 1);
        }
        setBlockedPositions(newBlockedPositions)
        break;
      case 'start':
        if (position === endPosition) return;
        setStartPosition(position);
        setRandomBlocks(position, endPosition, size);
        break;
      case 'end':
        if (position === startPosition) return;
        setEndPosition(position);
        setRandomBlocks(startPosition, position, size);
        break;
      default:
        break;
    }
  }

  const undoMovement = useCallback(() => {
    if (tokenPositions.length === 1) return
    setTokenPositions([...tokenPositions.slice(1)]);
  }, [tokenPositions])

  const handleEndGame = useCallback(() => {
    if (endPosition === tokenPositions[0]) {
      return changeSize(size + 1)
    } else if (limitMovements - tokenPositions.length === 0) {
      setTokenPositions([startPosition])
    }
  }, [changeSize, endPosition, limitMovements, size, startPosition, tokenPositions])

  const moveToken = useCallback((direction) => {
    let newPosition = tokenPositions[0];
    switch (direction) {
      case 'Up':
        if (tokenPositions[0] - size >= 0) newPosition -= size;
        break;
      case 'Down':
        if ((tokenPositions[0] + size) / size < size) newPosition += size;
        break;
      case 'Left':
        if ((tokenPositions[0] % size - 1) >= 0) newPosition -= 1;
        break;
      case 'Right':
        if (((tokenPositions[0] % size) + 1) < size) newPosition += 1;
        break;
      default:
        return;
    }
    if ((limitMovements - tokenPositions.length) > 0
      && endPosition !== tokenPositions[0]
      && newPosition !== tokenPositions[0]
      && !blockedPositions.includes(newPosition)) {
      const newTokenPositions = [...tokenPositions];
      newTokenPositions.unshift(newPosition);
      setTokenPositions(newTokenPositions);
    }
  }, [blockedPositions, endPosition, limitMovements, size, tokenPositions])

  const handleKeyDown = useCallback((event) => {
    const { key, path } = event;
    if ( path[0]?.toString()?.includes('Input') || edittingStatus !== false) return;
    if (key.includes('Arrow')) return moveToken(key.slice(5));
    switch (key) {
      case 'Backspace':
        undoMovement();
        break;
      case ' ':
        handleEndGame();
        break;
      case 'r':
        setTokenPositions([startPosition]);
        break;
      case 'R':
        setTokenPositions([startPosition]);
        break;
      default:
        return;
    }
  }, [edittingStatus, handleEndGame, moveToken, startPosition, undoMovement]);

  useEffect(() => {
    document.addEventListener('keydown', handleKeyDown);
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, [handleKeyDown])

  const handleChangeEditing = useCallback((type) => {
    if (edittingStatus !== type) return setEdittingStatus(type);
    setEdittingStatus(false);
    setTokenPositions([startPosition]);
  }, [edittingStatus, startPosition])

  const win = endPosition === tokenPositions[0];
  const lost = limitMovements - tokenPositions.length <= 0;
  const leftMoves = limitMovements - tokenPositions.length;
  return <Container>
    {edittingStatus !== true &&
      <div>
        {!edittingStatus ?
          <>
            <Row justifyContent='flex-end'>
              <CornerUpLeft cursor='pointer' data-testid='undo-icon' color='#FFFF07'
                onClick={undoMovement} />
              <RefreshCw cursor='pointer' data-testid='reload-icon' color='#FFFF07'
                onClick={() => setTokenPositions([startPosition])} />
              <Settings cursor='pointer' data-testid='settings-icon' color='#FFFF07'
                onClick={() => handleChangeEditing(true)} />
            </Row>
            <Token win={win} lost={lost} left={tokenPositions[0] % size}
              top={Math.floor(tokenPositions[0] / size)} />
          </>
          :
          <GameTypography data-testid='back-to-menu' onClick={() => handleChangeEditing(true)}>
            <Row alignItems='center'><ArrowLeft size={15} /> Menu </Row>
          </GameTypography>
        }
        <Board editting={edittingStatus} handleEndGame={handleEndGame} win={win} lost={lost} time={time}
          blockedPositions={blockedPositions} endPosition={endPosition} size={size}
          startPosition={startPosition} slotClicked={handleSlotClicked}></Board>
        {edittingStatus === 'random' &&
          <Row justifyContent='center'><RefreshCw data-testid='random-icon' color='#FFFF07' onClick={handleSetRandom} /></Row>
        }
        {!edittingStatus &&
          <div>
            <GameTypography color='white'>Moves Left {leftMoves < 0 ? 0 : leftMoves}</GameTypography>
            <Joystick moveToken={moveToken} />
          </div>
        }
      </div>
    }
    {edittingStatus === true &&
      <Menu handleMovementsChange={handleLimitMovementsChange} size={size} handleSizeChange={changeSize}
        movements={limitMovements} handleChangeEditing={handleChangeEditing} edittingStatus={edittingStatus} />}
  </Container>
}
