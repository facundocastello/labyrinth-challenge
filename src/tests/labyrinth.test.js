import React from 'react';
import Labyrinth from '../pages/Labyrinth';
import { create } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import { initSize, maxSize } from '../utils/constants';

const getBlocks = ( getByTestId, size ) => {
  const blocks = [];
  for ( let index = 0; index < size * size; index++ ) {
    const slot = getByTestId( `slot-${index}` );
    if ( slot.getAttribute( 'type' ) === 'blocked' ) blocks.push( index );
  }
  return blocks;
};

const selectMenuOption = ( getByTestId, id ) => {
  const settingsIcon = getByTestId( 'settings-icon' );
  fireEvent.click( settingsIcon );
  const action = getByTestId( id );
  fireEvent.click( action );
};

const cleanBlocks = ( getByTestId, size ) => {
  const blockedPositions = getBlocks( getByTestId, size );
  selectMenuOption( getByTestId, 'edit-block' );
  blockedPositions.forEach( blockPosition => {
    const slot = getByTestId( `slot-${blockPosition}` );
    fireEvent.click( slot );
  } );
};

describe( 'Labyrinth Block', () => {
  it( 'renders Labyrinth', () => {
    const tree = create( <Labyrinth /> ).toJSON();
    expect( tree ).toMatchSnapshot();
  } );

  it( 'removes blocks and expect to have blocksPositions empty', () => {
    const { getByTestId } = render( <Labyrinth /> );
    cleanBlocks( getByTestId, initSize );
    const blocks = [];
    for ( let index = 0; index < initSize * initSize; index++ ) {
      const slot = getByTestId( `slot-${index}` );
      if ( slot.getAttribute( 'type' ) === 'blocked' ) blocks.push( slot );
    }
    expect( blocks.length ).toBe( 0 );
  } );

  it( 'removes blocks, go to end and expect to see the you win message', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    cleanBlocks( getByTestId, initSize );
    fireEvent.click( getByTestId( 'back-to-menu' ));
    fireEvent.click( getByTestId( 'back-to-game' ));
    const rightKey = getByTestId( 'arrow-right' );
    const downKey = getByTestId( 'arrow-down' );
    for ( let index = 0; index < 4; index++ ) {
      fireEvent.click( rightKey );
      fireEvent.click( downKey );
    }
    expect( queryByText( 'You Win' )).not.toBeNull();
    const endGameButton = getByTestId( 'end-game' );
    fireEvent.click( endGameButton );
    expect( queryByText( 'You Win' )).toBeNull();
  } );

  it( 'removes blocks, move until you lose and expect to see the  you lose message', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    cleanBlocks( getByTestId, initSize );
    fireEvent.click( getByTestId( 'back-to-menu' ));
    fireEvent.click( getByTestId( 'back-to-game' ));
    const leftKey = getByTestId( 'arrow-left' );
    const upKey = getByTestId( 'arrow-up' );
    const rightKey = getByTestId( 'arrow-right' );
    const downKey = getByTestId( 'arrow-down' );
    for ( let index = 0; index < 3; index++ ) {
      fireEvent.click( leftKey );
      fireEvent.click( upKey );
      fireEvent.click( rightKey );
      fireEvent.click( downKey );
    }
    expect( queryByText( 'You Lose' )).not.toBeNull();
    const endGameButton = getByTestId( 'end-game' );
    fireEvent.click( endGameButton );
    expect( queryByText( 'You Lose' )).toBeNull();
  } );

  it( 'changes start position and check that blocks update', () => {
    const { getByTestId } = render( <Labyrinth /> );
    selectMenuOption( getByTestId, 'edit-start' );
    const slot = getByTestId( 'slot-1' );
    expect( slot.getAttribute( 'type' )).not.toBe( 'start' );
    fireEvent.click( slot );
    fireEvent.click( getByTestId( 'back-to-menu' ));
    fireEvent.click( getByTestId( 'back-to-game' ));
    expect( slot.getAttribute( 'type' )).toBe( 'start' );
  } );

  it( 'changes end position and check that blocks update', () => {
    const { getByTestId } = render( <Labyrinth /> );
    selectMenuOption( getByTestId, 'edit-end' );
    const slot = getByTestId( 'slot-2' );
    expect( slot.getAttribute( 'type' )).not.toBe( 'end' );
    fireEvent.click( slot );
    expect( slot.getAttribute( 'type' )).toBe( 'end' );
  } );

  it( 'gets "Moves Left", move token and expect "Moves Left" to be reduced by 1 and token to change position ', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    expect( queryByText( 'Moves Left 8' )).not.toBeNull();
    const slotRight = getByTestId( 'slot-1' );
    let key = slotRight.getAttribute( 'type' ) === 'blocked' ? 'down' : 'right';
    const keyButton = getByTestId( `arrow-${key}` );
    fireEvent.click( keyButton );
    expect( queryByText( 'Moves Left 7' )).not.toBeNull();
    const token = getByTestId( 'token' );
    expect( token ).toHaveStyleRule( key === 'down' ? 'top':'left', '37px' );
  } );

  it( 'goes to menu and goes back to game', () => {
    const { getByTestId } = render( <Labyrinth /> );
    selectMenuOption( getByTestId, 'back-to-game' );
    expect( getByTestId( 'board' )).not.toBeNull();
  } );

  it( 'changes movements and check that "Moves Left change"', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    const settingsIcon = getByTestId( 'settings-icon' );
    fireEvent.click( settingsIcon );
    const movements = getByTestId( 'Movements' );
    fireEvent.change( movements, { target: { value: 20 } } );
    const goBack = getByTestId( 'back-to-game' );
    fireEvent.click( goBack );
    expect( queryByText( 'Moves Left 20' )).not.toBeNull();
  } );

  it( 'changes size to be the max, win the game and expect the message to be Game Over ', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    const settingsIcon = getByTestId( 'settings-icon' );
    fireEvent.click( settingsIcon );
    const size = getByTestId( 'Size' );
    fireEvent.change( size, { target: { value: maxSize } } );
    fireEvent.click(getByTestId( 'edit-end' ));
    fireEvent.click(getByTestId( 'slot-1' ));
    fireEvent.click( getByTestId( 'back-to-menu' ));
    fireEvent.click( getByTestId( 'back-to-game' ));
    fireEvent.click( getByTestId( 'arrow-right' ) );
    expect( queryByText( 'Game Over' )).not.toBeNull();
  } );

  it( 'triggers undo and check that "Moves Left" changes', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    expect( queryByText( 'Moves Left 8' )).not.toBeNull();
    const slotRight = getByTestId( 'slot-1' );
    let key = slotRight.getAttribute( 'type' ) === 'blocked' ? 'down' : 'right';
    const keyButton = getByTestId( `arrow-${key}` );
    fireEvent.click( keyButton );
    expect( queryByText( 'Moves Left 7' )).not.toBeNull();
    const undoButton = getByTestId( 'undo-icon' );
    fireEvent.click( undoButton );
    expect( queryByText( 'Moves Left 8' )).not.toBeNull();
  } );

  it( 'triggers reload and check that "Moves Left" changes', () => {
    const { getByTestId, queryByText } = render( <Labyrinth /> );
    expect( queryByText( 'Moves Left 8' )).not.toBeNull();
    const slotRight = getByTestId( 'slot-1' );
    let key = slotRight.getAttribute( 'type' ) === 'blocked' ? 'down' : 'right';
    const keyButton = getByTestId( `arrow-${key}` );
    fireEvent.click( keyButton );
    expect( queryByText( 'Moves Left 7' )).not.toBeNull();
    const undoButton = getByTestId( 'reload-icon' );
    fireEvent.click( undoButton );
    expect( queryByText( 'Moves Left 8' )).not.toBeNull();
  } );
} );
