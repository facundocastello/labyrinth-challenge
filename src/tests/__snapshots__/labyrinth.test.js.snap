// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Labyrinth Block renders Labyrinth 1`] = `
.c6 {
  color: white;
  margin: 10px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  border: 1px solid black;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c0 {
  margin: 20px;
  padding: 20px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: black;
  border-radius: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c9 {
  background-color: white;
  margin: 1px;
  border-radius: 5px;
  cursor: pointer;
}

.c9:hover {
  opacity: .9;
}

.c7 {
  margin-top: 10px;
}

.c2 {
  position: relative;
  top: 5px;
  left: 5px;
  width: 0;
  height: 0;
  -webkit-transition: top .1s,left .1s;
  transition: top .1s,left .1s;
}

.c3 {
  width: 24px;
  height: 24px;
  -webkit-animation: heartbeat 1s infinite;
  animation: heartbeat 1s infinite;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      <svg
        cursor="pointer"
        data-testid="undo-icon"
        fill="none"
        height={24}
        onClick={[Function]}
        stroke="#FFFF07"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="9 14 4 9 9 4"
        />
        <path
          d="M20 20v-7a4 4 0 0 0-4-4H4"
        />
      </svg>
      <svg
        cursor="pointer"
        data-testid="reload-icon"
        fill="none"
        height={24}
        onClick={[Function]}
        stroke="#FFFF07"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="23 4 23 10 17 10"
        />
        <polyline
          points="1 20 1 14 7 14"
        />
        <path
          d="M3.51 9a9 9 0 0 1 14.85-3.36L23 10M1 14l4.64 4.36A9 9 0 0 0 20.49 15"
        />
      </svg>
      <svg
        cursor="pointer"
        data-testid="settings-icon"
        fill="none"
        height={24}
        onClick={[Function]}
        stroke="#FFFF07"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="3"
        />
        <path
          d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
        />
      </svg>
    </div>
    <div
      className="c2"
      data-testid="token"
    >
      <div
        className="c3"
      >
        <svg
          fill="none"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
      </div>
    </div>
    <div
      className="c4"
      data-testid="board"
    >
      <div
        className="c5"
      />
    </div>
    <div>
      <div
        className="c6"
        color="white"
      >
        Moves Left 
        8
      </div>
      <div
        className="c7"
      >
        <div
          className="c8"
          width="100%"
        >
          <div
            className="c9"
          >
            <svg
              data-testid="arrow-up"
              fill="none"
              height={45}
              onClick={[Function]}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={45}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="19"
                y2="5"
              />
              <polyline
                points="5 12 12 5 19 12"
              />
            </svg>
          </div>
        </div>
        <div
          className="c8"
          width="100%"
        >
          <div
            className="c9"
          >
            <svg
              data-testid="arrow-left"
              fill="none"
              height={45}
              onClick={[Function]}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={45}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="19"
                x2="5"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 19 5 12 12 5"
              />
            </svg>
          </div>
          <div
            className="c9"
          >
            <svg
              data-testid="arrow-down"
              fill="none"
              height={45}
              onClick={[Function]}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={45}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <polyline
                points="19 12 12 19 5 12"
              />
            </svg>
          </div>
          <div
            className="c9"
          >
            <svg
              data-testid="arrow-right"
              fill="none"
              height={45}
              onClick={[Function]}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={45}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 5 19 12 12 19"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
